File: python/koans/04_multiplex/test_multiplex.py
import threading

from koan_utils import (
    OverlapTracker,
    ThreadRunner,
    assert_blocks,
    assert_completes,
    eventually,
    jitter,
)

from multiplex import Multiplex


def test_never_more_than_n_inside():
    n = 4
    plex = Multiplex(n)
    tracker = OverlapTracker()
    runner = ThreadRunner()

    def worker():
        for _ in range(20):
            plex.enter()
            with tracker.section("room"):
                jitter(1.0)
            plex.exit()

    for _ in range(12):
        runner.spawn(worker)
    runner.join_all(timeout=30)
    assert tracker.max_concurrent("room") <= n, (
        f"room capacity {n} exceeded: saw {tracker.max_concurrent('room')}"
    )


def test_n_threads_can_be_inside_together():
    """A mutex in a trench coat must not pass: full concurrency is required."""
    n = 4
    plex = Multiplex(n)
    tracker = OverlapTracker()
    all_in = threading.Event()
    runner = ThreadRunner()

    def worker():
        plex.enter()
        tracker.enter("room")
        all_in.wait(5)  # linger until everyone is inside together
        tracker.exit("room")
        plex.exit()

    for _ in range(n):
        runner.spawn(worker)
    eventually(
        lambda: tracker.current("room") == n,
        timeout=5,
        msg=f"only {tracker.current('room')} of {n} threads made it inside "
        "concurrently — the multiplex over-serializes",
    )
    all_in.set()
    runner.join_all(timeout=5)


def test_thread_n_plus_1_blocks_then_proceeds_after_an_exit():
    n = 3
    plex = Multiplex(n)
    for _ in range(n):
        assert_completes(plex.enter, timeout=2)
    probe = assert_blocks(
        plex.enter, msg=f"entry {n + 1} must block while {n} are inside"
    )
    plex.exit()
    assert probe.wait(5), "a waiter should get the freed slot"
    for _ in range(n):
        plex.exit()


def test_capacity_one_degenerates_to_a_mutex():
    plex = Multiplex(1)
    assert_completes(plex.enter, timeout=2)
    probe = assert_blocks(plex.enter)
    plex.exit()
    assert probe.wait(5)
    plex.exit()


def test_stress_various_sizes():
    for n in (2, 5, 8):
        plex = Multiplex(n)
        tracker = OverlapTracker()
        runner = ThreadRunner()

        def worker(plex=plex, tracker=tracker):
            for _ in range(10):
                plex.enter()
                with tracker.section("room"):
                    jitter(0.5)
                plex.exit()

        for _ in range(n * 3):
            runner.spawn(worker)
        runner.join_all(timeout=30)
        assert tracker.max_concurrent("room") <= n

File: cpp/koans/04_multiplex/test_multiplex.cpp
#include "koan_test.hpp"
#include "multiplex.hpp"

#include <condition_variable>
#include <memory>
#include <thread>

using namespace koans;

KOAN_TEST(never_more_than_n_inside) {
    constexpr int n = 4;
    Multiplex plex(n);
    OverlapTracker tracker;
    ThreadRunner runner;
    for (int t = 0; t < 12; ++t) {
        runner.spawn([&] {
            for (int i = 0; i < 20; ++i) {
                plex.enter();
                tracker.enter("room");
                jitter(1);
                tracker.exit("room");
                plex.exit();
            }
        });
    }
    runner.join_all(30000ms);
    KOAN_ASSERT_MSG(tracker.max_concurrent("room") <= n,
                    "room capacity " + std::to_string(n) + " exceeded: saw " +
                        std::to_string(tracker.max_concurrent("room")));
}

// A mutex in a trench coat must not pass: full concurrency is required.
KOAN_TEST(n_threads_can_be_inside_together) {
    constexpr int n = 4;
    Multiplex plex(n);
    OverlapTracker tracker;
    std::atomic<bool> all_in{false};
    ThreadRunner runner;
    for (int t = 0; t < n; ++t) {
        runner.spawn([&] {
            plex.enter();
            tracker.enter("room");
            auto deadline = std::chrono::steady_clock::now() + 5s;
            while (!all_in.load() && std::chrono::steady_clock::now() < deadline)
                std::this_thread::sleep_for(1ms);  // linger until all inside
            tracker.exit("room");
            plex.exit();
        });
    }
    eventually([&] { return tracker.current("room") == n; }, 5000ms,
               "not all " + std::to_string(n) +
                   " threads made it inside concurrently — the multiplex "
                   "over-serializes");
    all_in.store(true);
    runner.join_all(5000ms);
}

KOAN_TEST(thread_n_plus_1_blocks_then_proceeds_after_an_exit) {
    constexpr int n = 3;
    auto plex = std::make_shared<Multiplex>(n);
    for (int i = 0; i < n; ++i)
        assert_completes([plex] { plex->enter(); }, 2000ms, "an in-capacity entry");
    auto probe = assert_blocks([plex] { plex->enter(); }, 300ms,
                               "entry " + std::to_string(n + 1) +
                                   " (room is full)");
    plex->exit();
    probe.assert_completed(5000ms, "the blocked entry after an exit");
    for (int i = 0; i < n; ++i) plex->exit();
}

KOAN_TEST(capacity_one_degenerates_to_a_mutex) {
    auto plex = std::make_shared<Multiplex>(1);
    assert_completes([plex] { plex->enter(); }, 2000ms, "first entry");
    auto probe = assert_blocks([plex] { plex->enter(); });
    plex->exit();
    probe.assert_completed(5000ms, "the second entry after an exit");
    plex->exit();
}

KOAN_TEST(stress_various_sizes) {
    for (int n : {2, 5, 8}) {
        Multiplex plex(n);
        OverlapTracker tracker;
        ThreadRunner runner;
        for (int t = 0; t < n * 3; ++t) {
            runner.spawn([&] {
                for (int i = 0; i < 10; ++i) {
                    plex.enter();
                    tracker.enter("room");
                    jitter(1);
                    tracker.exit("room");
                    plex.exit();
                }
            });
        }
        runner.join_all(30000ms);
        KOAN_ASSERT(tracker.max_concurrent("room") <= n);
    }
}
